hawtio-system/src/main/java/io/hawt/web/auth/ClientRouteRedirectFilter.java

```diff
@@ -0,0 +1,226 @@
+        }
+    }
+
+    @Override
+    public void doFilter(final ServletRequest request, final ServletResponse response, final FilterChain chain) throws IOException, ServletException {
+        LOG.trace("Applying {}", getClass().getSimpleName());
+
+        HttpServletRequest httpRequest = (HttpServletRequest) request;
+        HttpServletResponse httpResponse = (HttpServletResponse) response;
+        HttpSession session = httpRequest.getSession(false);
+
+        // TOCHECK: we may consider using this filter only for GET requests
+
+        // this is full path, which includes context path and path info
+        String requestURI = Strings.cleanPath(httpRequest.getRequestURI());
+        // this is a path without context path and should be everything after "/hawtio" (or "/actuator/hawtio")
+        String hawtioPath = requestURI.length() < baseFullPath.length() ? ""
+                : requestURI.substring(baseFullPath.length());
+        // this is a path after context path:
+        //  - everything after /hawtio for hawtio.war
+        //  - everything after /actuator/hawtio for default Spring Boot config. but also handles cases with custom
+        //    management base and/or mapping and if same port is used for main and management server, then dispatcher
+        //    servlet's base path is also handled (and this is the case why we CAN'T use req.getServletPath())
+        String path = requestURI.length() < contextPath.length() ? ""
+                : requestURI.substring(contextPath.length());
+
+        boolean loginPage = hawtioPath.startsWith("/login");
+
+        if (baseFullPath.equals(requestURI)) {
+            // explicitly change to "/" for Tomcat, so it's redirected if not authenticated
+            // for Spring Boot, "/actuator/hawtio" is changed to "/actuator/hawtio/", so we know it's top-level
+            path = "/".equals(basePath) ? "/" : basePath + "/";
+        }
+
+        LOG.debug("Check if path [{}] requires redirect", path);
+
+        // 0) whatever the configuration, accessing css, index.html, fonts, ..., should be handled normally
+        if (!loginPage && !isSecuredPath(path)) {
+            chain.doFilter(request, response);
+            return;
+        }
+
+        // "/login" is similar to URLs like "/connect/remote" or "/jmx", because it's client-side router URL
+        // but it shold be handled differently
+        // "/" (and even "") are also treated as client-side routes, because there's no path among
+        // io.hawt.web.auth.AuthenticationConfiguration.UNSECURED_PATHS that "/" startsWith()
+
+        // 1) no authentication or external authentication
+        if (!authConfiguration.isEnabled() || authConfiguration.isExternalAuthenticationEnabled()) {
+            if (loginPage) {
+                // /login should be redirected to "/", so user doesn't see login page blinking
+                redirector.doRedirect(httpRequest, httpResponse, "/");
```

Comment:
Forward to `/index.html` means simply - render html whatever's the incoming request. This is why we handle all client-side route URLs (like `/jmx`) like this - so user can refresh the page - server gets `/jmx`, responds with index.html content and user is happy.

Before this change, the above scenario was handled by Jimmi's error-page + error-code hack - you got index.html page, but with 404 and I didn't like it ;)

Redirect to `/` is kind of special. If you look closely now - when you're logged in (or authentication is disabled), browsing to `/login` gives you index.html, but react-router immediately analyzes window.location.href and sees `/login`, so client-routes to login page for a moment. I tried to avoid this. That's why the redirect.

Reply:
To be honest, I'm not yet fully convinced that we need this duplication of redirection in both hawtio/hawtio and hawtio-react. But let's accept it for now.

Reply:
We can treat it as duplication, but IMO we shouldn't rely on one side (the less secure one) to be the only performing proper navigation.
It's still _less_ duplication than before this PR ;) I could remove some Quarkus and SpringBoot specific redirect/forward mechanisms.